compiler/src/dotty/tools/dotc/transform/PostTyper.scala

```diff
@@ -279,6 +280,17 @@ class PostTyper extends MacroTransform with InfoTransformer { thisPhase =>
       if !tree.symbol.is(Package) then tree
       else errorTree(tree, em"${tree.symbol} cannot be used as a type")
 
+    // Cleans up retains annotations in inferred type trees. This is needed because
+    // during the typer, it is infeasible to correctly infer the capture sets in most
+    // cases, resulting ill-formed capture sets that could crash the pickler later on.
+    // See #20035.
+    private def cleanupRetainsAnnot(symbol: Symbol, tpt: Tree)(using Context): Tree =
+      tpt match
+        case tpt: InferredTypeTree if !symbol.allOverriddenSymbols.hasNext =>
```

Comment:
Would be good to explain the condition. Something like:

 // if there are overridden symbols, the annotation comes from an explicit type of the overridden symbol, and should be retained.